engine/access/rpc/backend/backend_network.go

```diff
@@ -94,26 +108,32 @@ func (b *backendNetwork) GetProtocolStateSnapshotByBlockID(_ context.Context, bl
 		return nil, status.Errorf(codes.Internal, "could not get header by blockID: %v", err)
 	}
 
-	snapshotByHeight := b.state.AtHeight(snapshotHeadByBlockId.Height)
-	snapshotHeadByHeight, err := snapshotByHeight.Head()
-	if err != nil {
-		if errors.Is(err, state.ErrUnknownSnapshotReference) {
-			return nil, status.Errorf(codes.InvalidArgument,
-				"failed to retrieve snapshot for block by height %d: block not finalized", snapshotHeadByBlockId.Height)
-		}
-
-		return nil, status.Errorf(codes.Internal, "failed to find snapshot: %v", err)
+	// Because there is no index from block ID to finalized height, we separately look up the finalized
+	// block ID by the height of the queried block, then compare the queried ID to the finalized ID.
+	// If they match, then the queried block must be finalized.
+	blockIDFinalizedAtHeight, err := b.headers.BlockIDByHeight(snapshotHeadByBlockId.Height)
+
```

Comment:
```suggestion
	blockIDFinalizedAtHeight, err := b.headers.BlockIDByHeight(snapshotHeadByBlockId.Height)
```
error assignment adjacent to checking block